### ISSUE TITLE
Auto-contrast leading to over saturated images

### ISSUE THREAD

**Issue by adosnsantos** [napari/napari#5608]
## 🐛 Bug

When viewing my fluorescence microscopy image in napari the auto contrast leads to significantly oversaturated images and I have to manually update the contrast limits range

Here is a link to my data in case this is useful

https://drive.google.com/file/d/1zDJQDCMBwFEixw_GtvmXrN0ELERhUyM4/view?usp=share_link

**Comment by Czaki** [comment]
Could you please add a screenshot with an improperly saturated image and a proper one? This will simplify understanding the source of the bug so should speedup fixing.

**Comment by alisterburt** [comment]
@Czaki I was with @adosnsantos when this came up and I think the issue is the same as https://forum.image.sc/t/contrast-of-inside-napari-cannot-be-adjusted-properly/55330/1

The image was a timelapse in an nd2 file loaded into napari with aicsimagio - I think this loads in as a dask array which may be the root cause here

**Comment by adosnsantos** [comment]
default:
<img width="600" alt="image" src="https://user-images.githubusercontent.com/127216746/223495804-9546cc21-c4cb-47e9-bdc9-6c2d58595c68.png">

after manually adjusting contrast:
<img width="600" alt="image" src="https://user-images.githubusercontent.com/127216746/223495995-63d69208-4b84-4ea3-8151-ced7ab372c4e.png">

the data is indeed loaded as a dask array of shape (13, 1, 5840, 5840)

**Comment by adosnsantos** [comment]
when extracting a slice from the dask array as a numpy array the contrast estimation still looks the same

**Comment by Czaki** [comment]
Ok. 
So the core problem is here 
https://github.com/napari/napari/blob/c77e001bd7ed21303aba7a547aafdf3c90b1fed6/napari/layers/utils/layer_utils.py#L209-L220

I do not know why @grlee77 decide to use only 64x64 pixels from the center of the image to calculate contrast limits. 

I do not know what the proper solution for this is. How better sample data? I have hope that someone could find a better solution here

**Comment by psobolewskiPhD** [comment]
Yeah I came to the same conclusion. I'm OK with the center slice, this makes sense most of the time I think, but 1e7 is ~3K by 3K, so 64x64 in the center is pretty specific.
I think it would be better to resample? `data[::10, ::10]` for example? Would be more representative.

**Comment by Czaki** [comment]
> I think it would be better to resample? `data[::10, ::10]` for example? Would be more representative.

But it will require loading all data. It is not a problem when data are in a numpy array, but it may be a problem when using dask or zarr backend. 

Btw. contrast limits should be a lazy attribute, not calculated in the constructor.

**Comment by psobolewskiPhD** [comment]
Sorry i wasn't clear, I meant using the existing code but instead of a slice that is `(c-n, c+n`), use a `step` like 10 or something.

**Comment by Czaki** [comment]
> Sorry i wasn't clear, I meant using the existing code but instead of a slice that is `(c-n, c+n`), use a `step` like 10 or something.

You are clear but see the image splitter in such a way:

```
┌────┬────┬────┐
│    │    │    │
│    │    │    │
├────┼────┼────┤
│    │    │    │
│    │    │    │
├────┼────┼────┤
│    │    │    │
│    │    │    │
└────┴────┴────┘
```

The current code will need to load only the middle tile and your proposition requires reading all tiles.

**Comment by psobolewskiPhD** [comment]
OK, makes sense I guess if the plane is really huge then loading all the chunks would be bad. 
But, if we're loading the middle chunk regardless, can we just use all of it? Chunk size will be bigger than 64x64 every time, right?

**Comment by Czaki** [comment]
It may be a good idea, but require deeper dive into backend-specific code to do this in a proper way. 

> Chunk size will be bigger than 64x64 every time, right?

If no then this should be a problem for the person who chunked data.

**Comment by psobolewskiPhD** [comment]
BTW, looks like it was actually @sofroniewn that implemented the 64x64 here: https://github.com/napari/napari/pull/2381

**Comment by Czaki** [comment]
> BTW, looks like it was actually @sofroniewn that implemented the 64x64 here: #2381

You are right. I believe too much to git blame.

**Comment by jni** [comment]
Thank you both for amazing sleuthing! 🕵️ 

A couple of ideas to bounce around:

- defer computing contrast limits until we have something loaded in a VisPy texture. Then we know that it's a "reasonable" size *and* loaded in RAM, so we can choose whatever sampling we want, even random sampling.
- give ourselves a time budget, say, 200ms. We grab increasingly large slices out of the middle of the 2D plane until we have used up the time budget, *then* we compute the contrast limits.

Dunno if either of these has merit given our architecture, but they are the two that come to mind.

**Comment by melonora** [comment]
Hi @adosnsantos,

I have opened a draft PR. I still need to add the higher than 1D case of RGB and fix some tests, but for your test data it is not correctly determining the contrast limits. Just to let you know, active work in progress:)

**Comment by xuefei-wang** [comment]
Hey! I came across the same issue when visualizing big images with multiple channels. Just want to provide one more data point here. This is what the one channel looks like: 

![Screenshot from 2024-04-03 21-56-40](https://github.com/napari/napari/assets/33240530/7c0eb495-eb86-4e87-8d62-38f49fe7c35b)

And this is what we expect it to look like: 

![Screenshot from 2024-04-03 21-57-40](https://github.com/napari/napari/assets/33240530/b59f15f6-4c09-4ec7-bbca-2c7a8426a5fb)

I appreciate the effort you're putting into the pull request, and I understand that it's a work in progress. Keep up the great work!

I was wondering if it might be beneficial to include a warning about the contrast limit calculation in both current and future versions. This would help users anticipate and address this issue if the auto calculation does not quite suit their needs.

**Comment by rossbar** [comment]
To second @xuefei-wang , the issue with default contrast limits is particularly pronounced when dealing with multiplexed tissue imaging (i.e. images with many channels). The [conditional on image size](https://github.com/napari/napari/blob/3cb47b5fc393427a09f3b98d641b3086ff93405e/napari/layers/utils/layer_utils.py#L246) essentially guarantees that this is the code path taken for multiplexed (i.e. >3 channels) images.

Re: minimal reproducers, there is a public repository of CODEX data on the [hubmap data portal](https://portal.hubmapconsortium.org/search?raw_dataset_type_keyword-assay_display_name_keyword[CODEX][0]=CODEX&raw_dataset_type_keyword-assay_display_name_keyword[CODEX][1]=CODEX%20%5BCytokit%20%2B%20SPRM%5D&entity_type[0]=Dataset). Here's [a link](https://portal.hubmapconsortium.org/browse/dataset/1bafd9eca3a630ebe851c8d2fe8cfa6a) to one of the smallest (~1GB) images I've found there [^1]. The image itself is downloadable under the `Files` tab at `ometiff-pyramids/pipeline_output/expr/reg001_expr.ome.tiff`.

Once the image is downloaded:

```python
>>> import tifffile as tff
>>> import numpy as np
>>> import napari
>>> napari.__version__
'0.4.19.post1'
>>> img = tff.imread("reg001_expr.ome.tif")
>>> img.shape
(33, 3457, 4603)
>>> img.nbytes
1050229686
>>> nim_left = napari.view_image(img, channel_axis=0)  # default contrast_limit
>>> cl = np.stack([np.min(img, axis=(1, 2)), np.max(img, axis=(1, 2))], axis=-1).tolist()
>>> nim_right = napari.view_image(img, channel_axis=0, contrast_limits=cl)
```

Here's a screenshot of the resulting images (default contrast limits on the left, manually computed on the right):

![multiplex_left-default_cl-right_explicit_cl](https://github.com/napari/napari/assets/1268991/fbba19f9-ec23-49b8-bb85-6b1a73715222)


It's definitely a tricky problem and I'm not sure that there is a single solution that strikes the right balance between performance and visual fidelity for all cases. I just wanted to draw extra attention to the multiplex case specifically, as I suspect this will be a very common problem with these types of data. At the very least, I'd +1 on a `UserWarning` when the sub-sampled contrast limits code path is hit.

[^1]: The issue can be demonstrated with essentially any one of these data sets, I just tried to find the smallest image to minimize download times.

**Comment by psobolewskiPhD** [comment]
This PR https://github.com/napari/napari/pull/6535 is some incremental progress on this front.
